### PR TITLE
Add a loss scale optimizer

### LINKED ISSUE #571: Port the loss scale optimizer to keras-core
We have been seeing some bad training runs in keras-nlp with a LLM running under mixed precision, and I think at least one of the missing pieces is scaling the loss. Scaling the loss up by 1k appeared to bring it in line with full precision.

https://github.com/keras-team/keras/blob/master/keras/mixed_precision/loss_scale_optimizer.py

This is a good explainer for why we need it ->
https://docs.nvidia.com/deeplearning/performance/mixed-precision-training/index.html#training

### PR DESCRIPTION
**Pull request by mattdangerw**
This is the big missing piece we need for feature parity when running mixed precision training compared to tf.keras.

Fixes #571

### CODE REVIEW

**Comment by mattdangerw** [comment]
A few points of awkwardness/discussion:

 1. `ops.cond` needs to be stateless for jax. Autoscaling has two cond branches with variables updates in all branches. To do this I overrode `stateless_apply` separately and had to do a lot of `StatelessScope`s. This feels very verbose and awkward, but I wasn't able to think of a great way around it. Suggestions welcome!
 2. We want `learning_rate` to proxy the inner optimizer learning rate. I overrode the `learning_rate` property to do so. But the base optimizer still has to be created with a learning rate, so I ended up just passing a zero valued variable which never gets used. This feels awkward and a bit confusing.
 3. Because the loss scale optimizer needs a variable to scale the loss, and we don't sync jax state except on epoch boundaries, I made a `scale_loss` and `stateless_scale_loss`. This is probably fine, just feels like a lot of code.

**Comment by mattdangerw** [comment]
Looks like there is some sort of device placement issue for `tensorflow` GPU that isn't picked up in our CPU only tests? Will poke around. Confirmed torch/jax are working.

File: keras_core/backend/tensorflow/optimizer.py [hunk 1]
```diff
@@ -86,8 +86,6 @@ def apply_grad_to_update_var(var, grad):
             return self.update_step(grad, var, learning_rate)
 
         for grad, var in grads_and_vars:
-            if isinstance(var, backend.Variable):
```
Comment (fchollet) [inline keras_core/backend/tensorflow/optimizer.py:89]: I suspect this may have been there for a reason. Why remove it?

File: keras_core/optimizers/loss_scale_optimizer.py [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+
+    def stateless_apply(self, optimizer_variables, grads, trainable_variables):
+        if not self.built:
+            raise ValueError(
+                f"To call `stateless_apply`, {self.__class__.__name__} "
+                "must be built (i.e. its variables must have been created). "
+                "You can build it via `optimizer.build(trainable_variables)`."
+            )
+
+        def handle_fininte_grads():
```
Comment (fchollet) [inline keras_core/optimizers/loss_scale_optimizer.py:93]: Typo

File: keras_core/trainers/trainer.py [hunk 3]
```diff
@@ -116,8 +118,21 @@ def compile(
...
+                scale the loss to prevent underflow.
         """
         self.optimizer = optimizers.get(optimizer)
+        if (
+            auto_scale_loss
+            and self.dtype_policy.name == "mixed_float16"
+            and self.optimizer
+            and not isinstance(self.optimizer, LossScaleOptimizer)
+        ):
+            self.optimizer = LossScaleOptimizer(
```
Comment (fchollet) [inline keras_core/trainers/trainer.py:133]: This should definitely be covered by tests

**Review by fchollet — COMMENTED** [review]
Thanks for the PR!

File: keras_core/trainers/trainer.py [hunk 4]
```diff
@@ -299,6 +314,22 @@ def metrics(self):
             total_loss = ops.sum(losses)
         return total_loss
 
+    def scale_loss(self, loss):
```
[the remark this answers is not in the document]
Comment (mattdangerw) [reply keras_core/trainers/trainer.py:317]: Yeah, was thinking the same while writing actually. I was following an [old comment from you](https://github.com/keras-team/keras-core/issues/571#issuecomment-1648838736), but let's try removing.

File: keras_core/backend/tensorflow/optimizer.py (hunk 1 above)
Comment (mattdangerw) [reply keras_core/backend/tensorflow/optimizer.py:89]: It was actually breaking `stateless_apply` on the tensorflow backend (which maybe we don't care about). The end error was attempt to run `eager_tensor.assign(x)`.

I could definitely just not test `stateless_apply` on tf, but I couldn't figure out why this was here, removing it fixed the issue, and broke no tests. Any ideas on why those two lines?

Comment (mattdangerw) [reply keras_core/backend/tensorflow/optimizer.py:89]: Alternately, I can just only test `stateless_apply` for jax, on only `apply` for tf/torch, that just seemed more janky.

**Comment by mattdangerw** [comment]
Addressed the initial round, though I may play with a test that deliberately triggers the underflow in trainer and asserts that variable updates appear. I don't think that should be too hard? But we will see.

File: keras_core/backend/tensorflow/optimizer.py (hunk 1 above)
Comment (fchollet) [reply keras_core/backend/tensorflow/optimizer.py:89]: SG

File: keras_core/optimizers/loss_scale_optimizer.py [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+from keras_core.api_export import keras_core_export
+from keras_core.optimizers import optimizer
+from keras_core.saving import serialization_lib
+from keras_core.utils import tracking
+
+
+@keras_core_export(
+    [
+        "keras_core.optimizers.LossScaleOptimizer",
+        "keras_core.mixed_precision.LossScaleOptimizer",
```
[the remark this answers is not in the document]
Comment (mattdangerw) [reply keras_core/optimizers/loss_scale_optimizer.py:13]: Yeah, IIUC, this was the only name it was ever exposed as, so if we want the backward compat we need this alias.

File: keras_core/optimizers/base_optimizer.py [hunk 6]
```diff
@@ -321,6 +327,22 @@ def stateless_apply(self, optimizer_variables, grads, trainable_variables):
...
+
+        Scales the loss before gradients are computed in a `train_step`. This
+        is primarily useful during mixed precision training to prevent numeric
+        underflow.
+        """
+        if self.loss_scale_factor is not None:
+            return loss * self.loss_scale_factor
+        return loss
+
+    def stateless_scale_loss(self, optimizer_variables, loss):
```
[the remark this answers is not in the document]
Comment (mattdangerw) [reply keras_core/optimizers/base_optimizer.py:341]: Good idea. I like the stateless scope. That keeps the weirdness isolated to jax.

Overall the jax trainer is definitely a little wonky (pushing state through `aux` for `jax.value_and_grad` is kinda bleh), but the real grossness is the jax implementation of the loss scale optimizer itself. Basically all jax control flow cannot be written then same was as tf/torch because a stateless scope is insufficient, you still have to return all state from each control flow callback. Not something to fix this PR, and I'm not sure we can really do anything about it, but it does break the idea of a backend agnostic graph of `ops`.

**Comment by mattdangerw** [comment]
Added a test end to end test only looking at variables updates across `fit()`, which will hopefully keep us from accidentally breaking this.

**Review by fchollet — APPROVED** [review]
Great work! LGTM
